R/simulate_infections.R

```diff
@@ -1,19 +1,205 @@
+#' @inheritParams stan_opts
+#' @importFrom lifecycle deprecate_warn
+#' @importFrom checkmate assert_data_frame assert_date assert_numeric
+#'   assert_subset
+#' @importFrom data.table data.table merge.data.table nafill rbindlist
+#' @return A data.table of simulated infections (variable `infections`) and
+#'   reported cases (variable `reported_cases`) by date.
+#' @author Sebastian Funk
 #' @export
-simulate_infections <- function(...) {
+#' @examples
+#' \donttest{
+#'   R <- data.frame(
+#'     date = seq.Date(as.Date("2023-01-01"), length.out = 14, by = "day"),
+#'     R = c(rep(1.2, 7), rep(0.8, 7))
+#'   )
+#'   sim <- simulate_infections(
+#'     R = R,
+#'     initial_infections = 100,
+#'     generation_time = generation_time_opts(
+#'       fix_dist(example_generation_time)
+#'     ),
+#'     delays = delay_opts(fix_dist(example_reporting_delay)),
+#'     obs = obs_opts(family = "poisson")
+#'   )
+#' }
+simulate_infections <- function(estimates, R, initial_infections,
+                                day_of_week_effect = NULL,
+                                generation_time = generation_time_opts(),
+                                delays = delay_opts(),
+                                truncation = trunc_opts(),
+                                obs = obs_opts(),
+                                CrIs = c(0.2, 0.5, 0.9),
+                                backend = "rstan",
+                                pop = 0, ...) {
+  ## deprecated usage
+  if (!missing(estimates)) {
     deprecate_warn(
       "2.0.0",
-      "simulate_infections()",
+      "simulate_infections(estimates)",
       "forecast_infections()",
-      "A new [simulate_infections()] function for simulating from given ",
-      "parameters is planned for implementation in the future."
+      details = paste0(
+        "This `estimates` option will be removed from [simulate_infections()] ",
+        "in version 2.1.0."
+      )
+    )
+    return(forecast_infections(estimates = estimates, ...))
+  }
+
+  ## check inputs
+  assert_data_frame(R, any.missing = FALSE)
+  assert_subset(colnames(R), c("date", "R"))
+  assert_date(R$date)
+  assert_numeric(R$R, lower = 0)
+  assert_numeric(initial_infections, lower = 0)
+  assert_numeric(day_of_week_effect, lower = 0, null.ok = TRUE)
+  assert_numeric(pop, lower = 0)
+  assert_class(delays, "delay_opts")
+  assert_class(obs, "obs_opts")
+  assert_class(generation_time, "generation_time_opts")
+
+  ## create R for all dates modelled
+  all_dates <- data.table(date = seq.Date(min(R$date), max(R$date), by = "day"))
+  R <- merge.data.table(all_dates, R, by = "date", all.x = TRUE)
+  R <- R[, R := nafill(R, type = "locf")]
+  ## remove any initial NAs
+  R <- R[!is.na(R)]
+
+  seeding_time <- get_seeding_time(delays, generation_time)
+  if (seeding_time > 1) {
+    ## estimate initial growth from initial reproduction number if seeding time
+    ## is greater than 1
+    initial_growth <- (R$R[1] - 1) / mean(generation_time)
```

Comment:
I don't think we are but we could be doing this in the inference model. In the Julia work Sam B implemented an improved version that doesn't use the mean approximation but I think this is likely good enough.

Reply:
See: https://github.com/CDCgov/Rt-without-renewal/blob/aae49d72646546c43ebec5b638ee712ea1310a69/EpiAware/src/utilities.jl#L159